### ISSUE TITLE
pkg_resources.ContextualVersionConflict

### ISSUE THREAD

**Issue by asottile** [PyCQA/flake8#195]
In GitLab by @olgabot on Nov 17, 2016, 08:05

Thanks for `flake8`, it's awesome! Keeps me in check when I'm changing too many things at once. I installed `flake8` using `pip` in this [Travis-CI](https://travis-ci.org/YeoLab/outrigger/jobs/175885323), using `pip install -r requirements.txt` and `flake8` is one of the requirements.

Oddly enough, all my tests run and my `make lint`, where the `lint` directive is below script passes, but ...

```make
lint:
	flake8 --exclude outrigger/external,doc outrigger
```

I also have a sanity check test to run my program on the command line and this is where the error occurs:

```
$ make lint
flake8 --exclude outrigger/external,doc outrigger
The command "make lint" exited with 0.
$ make arabdopsis
rm -rf *outrigger_output
outrigger index \
		--sj-out-tab outrigger/tests/data/arabdopsis/unprocessed/rna.chr4.subset.SJ.out.tab \
		--gtf outrigger/tests/data/arabdopsis/unprocessed/Arabidopsis_thaliana.TAIR10.31.chr4.subset.gtf \
		--min-reads 1 --n-jobs -1 \
		--output arabdopsis_outrigger_output
Traceback (most recent call last):
  File "/home/travis/miniconda/envs/testenv/bin/outrigger", line 6, in <module>
    from pkg_resources import load_entry_point
  File "/home/travis/miniconda/envs/testenv/lib/python2.7/site-packages/setuptools-27.2.0-py2.7.egg/pkg_resources/__init__.py", line 2985, in <module>
  File "/home/travis/miniconda/envs/testenv/lib/python2.7/site-packages/setuptools-27.2.0-py2.7.egg/pkg_resources/__init__.py", line 2971, in _call_aside
  File "/home/travis/miniconda/envs/testenv/lib/python2.7/site-packages/setuptools-27.2.0-py2.7.egg/pkg_resources/__init__.py", line 2998, in _initialize_master_working_set
  File "/home/travis/miniconda/envs/testenv/lib/python2.7/site-packages/setuptools-27.2.0-py2.7.egg/pkg_resources/__init__.py", line 662, in _build_master
  File "/home/travis/miniconda/envs/testenv/lib/python2.7/site-packages/setuptools-27.2.0-py2.7.egg/pkg_resources/__init__.py", line 675, in _build_from_requirements
  File "/home/travis/miniconda/envs/testenv/lib/python2.7/site-packages/setuptools-27.2.0-py2.7.egg/pkg_resources/__init__.py", line 859, in resolve
pkg_resources.ContextualVersionConflict: (pyflakes 1.3.0 (/home/travis/miniconda/envs/testenv/lib/python2.7/site-packages), Requirement.parse('pyflakes<1.1,>=0.8.1'), set(['flake8']))
make: *** [arabdopsis] Error 1
```

I don't expect any `flake8` error when I run this code because `flake8` isn't imported anywhere in my program. I only use it on the command line for lint checking. Do you think something could be happening because `make lint` occurs right before `make arabdopsis`?

Thank you!
Olga

**Comment by asottile** [comment]
In GitLab by @sigmavirus24 on Nov 17, 2016, 10:09

Without knowing what those make targets are doing, I can't provide you a good answer. What appears to happen (to me) is that something is changing the version of Flake8 you are relying on but not forcing it to change its dependencies as well. What that means is, something is forcibly downgrading Flake8 but not it's dependencies so when setuptools starts scanning entry points, it notes this version conflict and raises the exception, even if not using Flake8.

This is *not* a bug in Flake8 nor do I consider it one in setuptools. You should figure out what's forcing the downgrade of flake8 though.

**Comment by asottile** [comment]
In GitLab by @sigmavirus24 on Nov 17, 2016, 10:18

So [your project](https://github.com/YeoLab/outrigger) has conda_requirements and requirements.txt. In conda_requirements, you have `flake8` with no bound. The same is true for your `requirements.txt`. The difference that seems to agree with your problem is that you have `pyflakes` specified in your requirements.txt. So while conda installs: [flake8 2.5.1](https://travis-ci.org/YeoLab/outrigger/jobs/175885323#L349) you also manage to get too new a version of [pyflakes](https://travis-ci.org/YeoLab/outrigger/jobs/175885323#L395).

I don't know why this happens or how flake8 runs when it runs, but you've found a very interesting problem. If you can reproduce this against flake8 3.2, I'll look into how to fix this more closely, but 2.5.1 is no longer supported.

**Comment by asottile** [comment]
In GitLab by @olgabot on Nov 18, 2016, 13:16

FWIW: My solution is to remove flake8 from my installation requirements and
test linting in a different environment
